log-levels.md

```diff
@@ -0,0 +1,203 @@
+
+#### Logging with Correlation IDs / Trace IDs
+
+A very common pattern is to log messages with some "correlation id". The best approach in general here is to use a `LoggingContext` from [swift-distributed-tracing](https://github.com/apple/swift-distributed-tracing) as then your library will be able to be traced and used with correlation contexts regardless what tracing system the end-user is using (e.g. open telemetry, zipkin, xray etc.) The concept though can be explained well with just a manually logged "requestID," which we'll explain below.
+
+Consider an HTTP client as an example of a library that has a lot of metadata about some request, perhas something like this:
+
+```swift
+log.trace("Received response", metadata: [
+   "id": "...",
+   "peer.host": "...",
+   "payload.size": "...",
+   "headers": "...",
+   "responseCode": "...",
+   "responseCode.text": "...",
+])
+```
+
+The exact metadata does not matter, they're just some placeholder in this example. What matters is that there's "a lot of it".
+
+Now, we would like to avoid logging _all_ this information in every single log statement. Instead, we are able to just repeatedly log the `"id"` metadata, like this:
+
+```swift
+// ... 
+log.trace("Somehing something...", metadata: ["id": "..."])
+log.trace("Finished streaming response", metadata: ["id": "..."]) // good, the same ID is propagated
+```
+
+Thanks to the correlation ID (or a tracing provided ID, in which case we'd log as `context.log.trace("...")` as the ID is propagated automatically), in each following log statement after the initial log statement we're able to correlate all those log statements and know that this `"Finished streaming response"` message was about a response with a `responseCode` that we're able to look up from the `"Received response"` log message.
+
+This pattern is somewhat advanced and may not always be the right approach, but consider it in high performance code where logging the same information repeatedly can be too costly.
+
+##### Things to avoid with Correlation ID logging
+
+When logging with correlation contexts make sure to never "drop the ID", it is easiest to get this right when using distributed tracing's LoggingContext since propagating it ensures the carrying of identifiers, however the same applies to any kind of correlation identifier.
+
+Specifically, avoid situations like these:
+
+```swift
+debug: conection established [connection-id: 7]
+debug: connection closed unexpectedly [error: foobar] // BAD, the connection-id was dropped
+```
+
+On the second line, we don't know which connection had the error since the `connection-id` was dropped. Make sure to audit your logging code to ensure all relevant log statements carry neccessary correlation identifiers.
+
+#### Structured Logging (Semantic Logging)
```

Comment:
I think we should move this above "Logging with Correlation IDs / Trace IDs" -- since its more common

Reply:
Okey, sounds good